### PR TITLE
[#162031] [Shared dev] Duration based pricing - Move form inputs to Pricing tab

### PR DESCRIPTION
**Pull request by LeticiaErrandonea**
# Release Notes
Instruments with Duration pricing mode have new fields in the Pricing rules table. There Stepped billing rates can be set.

Adjustment input now has a `-` before it, so we're explicit about it being a discount.

Pending:
- Validate that Stepped billing rate <= base rate
- Validate that Stepped billing adjustment <= base rate
- Improve error messages

# Screenshot
image

# Additional Context

Optional. Feel free to add/modify additional headers as appropriate, e.g. "Refactorings", "Concerns".

### CODE REVIEW

File: app/services/price_policy_updater.rb [hunk 1]
```diff
@@ -39,10 +40,27 @@ def assign_attributes
...
   end
 
+  def save_product
+    keys_from_built_rate_starts = []
+
+    rate_starts_params = permitted_product_params[:rate_starts_attributes].clone
+    rate_starts_params.values.each_with_index do |dr, index|
+      if dr[:min_duration].blank?
+        if dr[:id].present?
+          rate_starts_params["#{index}"].merge! ({ _destroy: '1' })
```
Comment (giladshanan) [inline app/services/price_policy_updater.rb:54]: Curious how the various validations work together - for example:
- can you have a rate start with no duration rates?
- if you have a rate start with duration rates, and you remove the rate start, what happens?

File: app/services/price_policy_updater.rb [hunk 2]
```diff
@@ -39,10 +40,27 @@ def assign_attributes
...
+        else
+          keys_from_built_rate_starts << "#{index}"
+        end
+      end
+    end
+
+    @product.update({ rate_starts_attributes: rate_starts_params.except(*keys_from_built_rate_starts) })
```
Comment (giladshanan) [inline app/services/price_policy_updater.rb:61]: A note for later, if we have time: I think we should have a separate method that assembles the appropriate data from params.  Then we should be able to pass it over to the price_policy:
```
def price_policy_attributes
  price_group_attributes(price_policy.price_group).merge(product_attributes)
end

def product_attributes
 keys_from_built_rate_starts = []
  ...
  { rate_starts_attributes: rate_starts_params.except(*keys_from_built_rate_starts) }
end

price_policy.assign_attributes(price_policy_attributes)

```

File: app/services/price_policy_updater.rb (hunk 1 above)
Comment (LeticiaErrandonea) [reply app/services/price_policy_updater.rb:54]: > can you have a rate start with no duration rates?
Pushed this to when I add Duration rates. I was planning on validating that and returning an error. Not sure where I'd place it, though. Maybe in the Updater itself?

> if you have a rate start with duration rates, and you remove the rate start, what happens?
Should be destroyed

Comment (giladshanan) [reply app/services/price_policy_updater.rb:54]: @LeticiaErrandonea for the second case

> if you have a rate start with duration rates, and you remove the rate start, what happens?

I think we should either show an error or use JS to display a confirmation modal

Comment (LeticiaErrandonea) [reply app/services/price_policy_updater.rb:54]: So if a user wants to delete the Step, they need to delete all duration rates? I think that may be a lot of work for the user, but I don't know how many price groups is common to have

File: app/models/duration_rate.rb [hunk 3]
```diff
@@ -2,10 +2,26 @@
...
-  validates :min_duration, numericality: { greater_than_or_equal_to: 0, allow_blank: true }
-  validates :rate, presence: true
+  validate :rate_or_subsidy
   validates :rate, numericality: { greater_than_or_equal_to: 0, allow_blank: true }
+  validates :subsidy, numericality: { greater_than: 0, allow_blank: true }
+
+  # TODO: Validate:
+  # - Rate is <= base rate (if Price group is the “base” group or external).
+  # - Subsidy is <= base rate (if Price group is internal but not the “base”).
+  # For this it looks like we'll need to link Rate Start to Price Policy, instead of Instrument.
```
Comment (giladshanan) [inline app/models/duration_rate.rb:15]: Totally in support of implementing this later, but wanted to share my thoughts.  Each product should only have one _current_ price policy per price group - and in the context of the controller I think we should have all the relevant data by way of the price policy object. The price policy knows its price group and the price groups knows the duration rate.  I don't think we need rate start here, do we?

File: app/models/duration_rate.rb [hunk 4]
```diff
@@ -2,10 +2,26 @@
...
+  # In this context we only have a Price Group (which can have N Price policies) and an Instrument (which can have N Price policies).
+  # We need a Price Policy as that's where we store the usage_rate.
+  # Also, makes sense that Steps for an Instrument can change in time but we may want to keep the historic data.
```
Comment (giladshanan) [inline app/models/duration_rate.rb:18]: We're going to log changes to the Event Log, any other historic tracking is out of scope for now.

File: app/services/price_policy_updater.rb [hunk 5]
```diff
@@ -39,10 +44,47 @@ def assign_attributes
 
   def save
     ActiveRecord::Base.transaction do
-      @price_policies.all?(&:save) || raise(ActiveRecord::Rollback)
+      (assign_attributes && @price_policies.all?(&:save)) || raise(ActiveRecord::Rollback)
+    end
+  end
+
+  def assign_attributes_for_stepped_billing
```
Comment (giladshanan) [inline app/services/price_policy_updater.rb:51]: We should consider how to refactor this in the next sprint

**Review by giladshanan — COMMENTED** [review]
This is looking pretty good to me - are you planning to add specs here?  I would vote for doing logging in the next sprint unless you already have work in progress for that

File: app/services/price_policy_updater.rb [hunk 6]
```diff
@@ -39,10 +44,53 @@ def assign_attributes
 
   def save
     ActiveRecord::Base.transaction do
-      @price_policies.all?(&:save) || raise(ActiveRecord::Rollback)
+      (assign_attributes && @price_policies.all?(&:save)) || raise(ActiveRecord::Rollback)
```
Comment (giladshanan) [inline app/services/price_policy_updater.rb:47]: @LeticiaErrandonea should this change be dependent on the product pricing mode?

Comment (giladshanan) [reply app/services/price_policy_updater.rb:47]: Actually, I think this method is only called on line 33, so we should revert this change

File: app/models/duration_rate.rb (hunk 3 above)
Comment (LeticiaErrandonea) [reply app/models/duration_rate.rb:15]: I understand what you mean.

I tried to rise 2 different questions here:
1. Business: does it make sense to have only one set of duration rates per instrument? Should I stick to a set of intervals for all my price policies? I definitely lack knowledge here, but I'm wondering whether it makes sense to say that for one price policy my intervals were 1-3, 3-5, 5+ hours, and then for my new Price policy I want them to be 2-4, 4-5, 5+
2. How to get the base rate. Now I understand I could just look for the current Price policy for the instrument and that's it 🙂

File: app/services/price_policy_updater.rb (hunk 5 above)
Comment (LeticiaErrandonea) [reply app/services/price_policy_updater.rb:51]: Yes, 100%!

**Review by giladshanan — APPROVED** [review]
Pending CI

**Pull request merged**